.div_inline_block <- function(width, content) { shiny::div(style = paste0("display: inline-block; vertical-align: top; width: ", width, "cm"), content)}
.div_inline_br_block <- function(width) { .div_inline_block(width, shiny::br()) }
.prettySwitch <- function(id, name) { shinyWidgets::prettySwitch(id, name, fill = TRUE, status = "primary", value = FALSE) }

.plot_sidebar_layout <- function() {
    shiny::sidebarLayout(
        # Sidebar panel - a DT table listing the direct outliers
        shiny::sidebarPanel(
            # Select phenotype
            shiny::selectInput(
                inputId = "select_phenotype",
                label = "Select phenotype:",
                choices = c("No phenotype selected")
            ),
            # Input - select one or multiple rows
            shiny::radioButtons(
                inputId = "select_row_type",
                label = "Select single or multiple rows:",
                choices = c(Single = "single", Multiple = "multiple"),
                selected = "single",
                inline = TRUE
            ),
            shiny::div(DT::DTOutput("outliers_table"), style = "font-size:70%")
        ),

        # Main panel - various plots shown in sub-panels
        shiny::mainPanel(
            shiny::tabsetPanel(
                type = "tabs",
                .gwes_plot_panel(),
                .tree_plot_panel(),
                .circular_plot_panel()
            )
        )
    )
}

.about_panel <- function() {
    shiny::tabPanel(title = "About",
        shiny::h2("GWESalyzer - a genome-wide epistasis analyzer."),
        shiny::p("GWESalyzer is an interactive visualization tool designed to be used in conjunction with ",
        shiny::HTML('<a href="https://github.com/santeripuranen/SpydrPick" target="_blank">SpydrPick</a>.'),
        "GWESalyzer relies heavily on the Shiny and Vega libraries to deploy it as a web application with multiple plot types ",
        "to visualize the genomic data. GWESalyzer operates on the outliers file generated by SpydrPick ",
        "to produce a GWES Manhattan plot. Two additional plots can be generated, namely a phylogenetic tree ",
        "plot by providing either a Newick or a Nexus file in addition to a FASTA and a loci file, ",
        "and a highly interactive circular chromosome plot by providing a GFF3 file."),
        style = "font-size: 16px; word-wrap: break-word; width: 800px")
}

# Define UI for application
.ui <- shiny::fluidPage(
    shinyjs::useShinyjs(),

    # Page with several tabs
    shiny::navbarPage(title = "GWESalyzer",
       id = "tabs",
       .about_panel(),
       shiny::tabPanel(title = "Upload data", .data_sidebar_layout()),
       shiny::tabPanel(title = "Analyse SpydrPick output", .plot_sidebar_layout())
    )
)
