.ui_about_tab_panel <- function() {
    shiny::tabPanel(
        title = "About",
        shiny::h2("GWES-Explorer - a genome-wide epistasis analyzer."),
        shiny::p("GWES-Explorer is an interactive visualization tool designed to be used in conjunction with ",
                 shiny::HTML('<a href="https://github.com/santeripuranen/SpydrPick" target="_blank">SpydrPick</a>.'),
                 "GWES-Explorer relies heavily on the Shiny and Vega libraries to deploy it as a web application with ",
                 "multiple plot types to visualize the genomic data. GWES-Explorer operates on the outliers file ",
                 "generated by SpydrPick to produce a GWES Manhattan plot. Two additional plots can be generated, ",
                 "namely a phylogenetic tree plot by providing either a Newick or a Nexus file in addition to a FASTA ",
                 "and a loci file, and a highly interactive circular chromosome plot by providing a GFF3 file.",
                 shiny::br(),
                 shiny::br(),
                 "GitHub: ",
                 shiny::HTML('<a href="https://github.com/jurikuronen/GWES-Explorer" target="_blank">',
                             'https://github.com/jurikuronen/GWES-Explorer</a>')
        ),
        style = paste("font-size: 16px;",
                      "word-wrap: break-word;",
                      "width: 800px")
    )
}

.ui_upload_data_sidebar_panel <- function() {
    shiny::sidebarPanel(
        shiny::p("All plots require the SpydrPick outliers file."),
        shiny::uiOutput("outliers_file_input"),
        shiny::hr(),

        shiny::p("Tree plot requires tree, fasta and loci files. Optionally, you can upload a phenotypic data file."),
        shiny::uiOutput("tree_file_input"),
        shiny::uiOutput("fasta_file_input"),
        shiny::uiOutput("loci_file_input"),
        shiny::uiOutput("phenotype_file_input"),
        shiny::hr(),

        shiny::p("Circular plot requires a gff file."),
        shiny::uiOutput("gff_file_input"),
        shiny::hr(),

        shiny::actionButton(inputId = "read_data_button",
                            label = "Read in data"),
        .div_inline_br_block(3),
        shiny::actionButton(inputId = "read_example_data_button",
                            label = "Load example data (Maela)"),
        .div_inline_br_block(3),
        .div_inline_block(4.5, shiny::p(id = "reading_data_div", shiny::textOutput("data_loaded")))
    )
}

.ui_upload_data_main_panel <- function() {
    shiny::mainPanel(
        shiny::h3("Data formats"),
        shiny::br(),
        shiny::h4("SpydrPick outliers file (.outliers, .txt)"),
        shiny::p("The outliers file obtained with SpydrPick should be a space-delimited no-headers file with at least ",
                 "the following columns:"),
        shiny::code("Pos_1", shiny::HTML("&nbsp;"),
                    "Pos_2", shiny::HTML("&nbsp;"),
                    "Distance", shiny::HTML("&nbsp;"),
                    "Direct", shiny::HTML("&nbsp;"),
                    "MI", shiny::HTML("&nbsp;"),
                    "MI_wogaps"),
        shiny::br(), shiny::br(),
        shiny::p("Above, ", shiny::code("Pos_1"), " and ", shiny::code("Pos_2"), " refer to a pair of positions in ",
                 "the output that are ", shiny::code("Distance")," base pairs apart and whose MI score is",
                 shiny::code("MI"), " and MI score without gaps is ", shiny::code("MI_wogaps"), ". The ",
                 shiny::code("Direct"), " column is a 1/0 Boolean value from the ARACNE filtering step, where 0 ",
                 "indicates being filtered out by ARACNE."),
        shiny::br(),
        shiny::strong("All plots provided by GWES-Explorer require the outliers file."),
        shiny::br(), shiny::br(),
        shiny::h4("Phylogenetic tree files"),
        shiny::p("The phylogenetic tree plot requires a Newick or Nexus tree file, a fasta and a loci file.",
                 "Optionally, you can upload a phenotypic data file."),
        shiny::br(),
        shiny::h4("GFF3 file for the circular plot"),
        shiny::p("The circular plot requires a GFF3 file with the fields"),
        shiny::code("type", shiny::HTML("&nbsp;"),
                    "start", shiny::HTML("&nbsp;"),
                    "end", shiny::HTML("&nbsp;"),
                    "attributes"),
        shiny::br(), shiny::br(),
        shiny::p("defined. GWES-Explorer primarily looks for \"gene\" in the ", shiny::code("type"), "field and ",
                 "secondarily \"CDS\". Further, the ", shiny::code("attributes"), " field should have a ",
                 shiny::code("Name"), " tag present.")
    )
}

.ui_upload_data_tab_panel <- function() {
    shiny::tabPanel(
        title = "Upload data",
        shiny::sidebarLayout(
            .ui_upload_data_sidebar_panel(),
            .ui_upload_data_main_panel()
        )
    )
}

.ui_analyse_data_tab_panel <- function() {
    shiny::tabPanel(
        title = "Analyse data",
        shiny::sidebarLayout(
            # Sidebar panel - a DT table listing the direct outliers.
            shiny::sidebarPanel(
                # Drop-down list for selecting phenotype.
                shiny::selectInput(
                    inputId = "select_phenotype",
                    label = "Select phenotype:",
                    choices = c("No phenotype selected")
                ),
                # Setting to allow selecting multiple outlier rows.
                shiny::radioButtons(
                    inputId = "select_row_type",
                    label = "Select single or multiple rows:",
                    choices = c(Single = "single",
                                Multiple = "multiple"),
                    selected = "single",
                    inline = TRUE
                ),
                # Display outliers table.
                shiny::div(DT::DTOutput("outliers_table"),
                           style = "font-size: 70%")
            ),

            # Main panel with various plot tabs available.
            shiny::mainPanel(
                shiny::tabsetPanel(
                    type = "tabs",
                    .gwes_plot_panel(),
                    .tree_plot_panel(),
                    .circular_plot_panel()
                )
            )
        )
    )
}
